nexus/internal/filetransfer/file_transfer_manager.go

```diff
@@ -0,0 +1,223 @@
+
+	"github.com/wandb/wandb/nexus/pkg/observability"
+)
+
+type Storage int
+
+const (
+	bufferSize              = 32
+	defaultConcurrencyLimit = 128
+)
+
+type FileTransfer interface {
+	Upload(task *UploadTask) error
+	Download(task *DownloadTask) error
+}
+
+// FileTransferManager handles the upload/download of files
+type FileTransferManager struct {
+	// inChan is the channel for incoming messages
+	inChan chan interface{}
+
+	// fileTransfer is the uploader/downloader
+	// todo: make this a map of uploaders for different destination storage types
+	fileTransfer FileTransfer
+
+	// semaphore is the semaphore for limiting concurrency
+	semaphore chan struct{}
+
+	// fileCounts is the file counts for the file transfer
+	fileCounts *service.FileCounts
+
+	// fileCountsChan is the channel used to count files uploaded/downloaded by type
+	fileCountsChan chan FileType
+
+	// wgfc is the wait group for the file counts channel ops
+	wgfc *sync.WaitGroup
+
+	// settings is the settings for the file transfer
+	settings *service.Settings
+
+	// logger is the logger for the file transfer
+	logger *observability.NexusLogger
+
+	// wg is the wait group
+	wg *sync.WaitGroup
+}
+
+type FileTransferManagerOption func(fm *FileTransferManager)
+
+func WithLogger(logger *observability.NexusLogger) FileTransferManagerOption {
+	return func(fm *FileTransferManager) {
+		fm.logger = logger
+	}
+}
+
+func WithSettings(settings *service.Settings) FileTransferManagerOption {
+	return func(fm *FileTransferManager) {
+		fm.settings = settings
+	}
+}
+
+func WithFileTransfer(fileTransfer FileTransfer) FileTransferManagerOption {
+	return func(fm *FileTransferManager) {
+		fm.fileTransfer = fileTransfer
+	}
+}
+
+func (fm *FileTransferManager) fileCount() {
+	for fileType := range fm.fileCountsChan {
+		switch fileType {
+		case WandbFile:
+			fm.fileCounts.WandbCount++
+		case MediaFile:
+			fm.fileCounts.MediaCount++
+		case ArtifactFile:
+			fm.fileCounts.ArtifactCount++
+		default:
+			fm.fileCounts.OtherCount++
+		}
+	}
+	fm.wgfc.Done()
+}
+
+func NewUploadManager(opts ...FileTransferManagerOption) *FileTransferManager {
+
+	fm := FileTransferManager{
+		inChan:         make(chan interface{}, bufferSize),
+		fileCounts:     &service.FileCounts{},
+		fileCountsChan: make(chan FileType, bufferSize),
+		wgfc:           &sync.WaitGroup{},
+		wg:             &sync.WaitGroup{},
+		semaphore:      make(chan struct{}, defaultConcurrencyLimit),
+	}
+
+	for _, opt := range opts {
+		opt(&fm)
+	}
+
+	fm.wgfc.Add(1)
+	go fm.fileCount()
+
+	return &fm
+}
+
+// Start is the main loop for the fileTransfer
+func (fm *FileTransferManager) Start() {
+	fm.wg.Add(1)
+	go func() {
+		for task := range fm.inChan {
+			// add a task to the wait group
+			fm.wg.Add(1)
+			fm.logger.Debug("fileTransfer: got task", "task", task)
+			// spin up a goroutine per task
+			go func(task interface{}) {
+				// Acquire the semaphore
+				fm.semaphore <- struct{}{}
+				switch t := task.(type) {
```

Comment:
The two branches here are identical, why do we need the switch?

Reply:
Same as below. Without the switch, Go doesn't know how to resolve the `interface{}` - do you have any other suggestions on how to handle this though?